hw/mcu/nxp/kinetis/src/hal_timer.c

```diff
@@ -0,0 +1,431 @@
+    uint16_t overflow; /* MSB of the current counter is handled manually */
+    TAILQ_HEAD(hal_timer_qhead, hal_timer) hal_timer_q;
+};
+
+#if MYNEWT_VAL(TIMER_0)
+struct kinetis_hal_tmr kinetis_tmr0 = {
+    .base = LPTMR0,
+};
+#endif
+#if MYNEWT_VAL(TIMER_1)
+struct kinetis_hal_tmr kinetis_tmr1 = {
+    .base = LPTMR1,
+};
+#endif
+
+static const IRQn_Type kinetis_timer_irqs[] = LPTMR_IRQS;
+
+static const struct kinetis_hal_tmr *kinetis_timers[KINETIS_TIMERS_MAX] = {
+#if MYNEWT_VAL(TIMER_0)
+    &kinetis_tmr0,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_1)
+    &kinetis_tmr1,
+#else
+    NULL,
+#endif
+};
+
+static uint32_t
+kinetis_tmr_read(const struct kinetis_hal_tmr *tmr)
+{
+    return ((uint32_t)(tmr->overflow << 16) + LPTMR_GetCurrentTimerCount(tmr->base));
+}
+
+static void
+kinetis_tmr_set_period(struct kinetis_hal_tmr *tmr, uint32_t tick)
+{
+    /* CMR cannot be zero, so add one cycle */
+    if ((tick & 0xffff) == 0) {
+        tick++;
+    }
+    /* XXX will only set the lower 16-bits */
+    LPTMR_SetTimerPeriod(tmr->base, tick);
+}
+
+static uint32_t
+kinetis_tmr_get_freq(const struct kinetis_hal_tmr *tmr)
+{
+    uint32_t psr;
+    uint32_t freq;
+
+    psr = tmr->base->PSR;
+    switch (psr & LPTMR_PSR_PCS_MASK) {
+    case 0:
+        freq = MCGIRCLK_HZ;
+        /* if the clock divisor is not bypassed, apply the divider */
+        if (!(psr & LPTMR_PSR_PBYP_MASK)) {
+            freq >>= ((psr & LPTMR_PSR_PRESCALE_MASK) >> LPTMR_PSR_PRESCALE_SHIFT) + 1;
+        }
+        break;
+    default:
+        /* Only MCGIRCLK is supported at the moment */
+        assert(0);
+    }
+
+    return freq;
+}
+
+static void
+kinetis_tmr_config_freq(lptmr_config_t *config, uint32_t freq_hz)
+{
+    uint32_t tmr_freq;
+    uint8_t prescaler;
+
+    config->prescalerClockSource = kLPTMR_PrescalerClock_0;
+
+    tmr_freq = MCGIRCLK_HZ;
+    prescaler = 0xff; /* overflow on purpose */
+    while (freq_hz < tmr_freq) {
+        tmr_freq /= 2;
+        prescaler++;
+        /* avoid a second overflow */
+        if (prescaler == 0xff) {
+            break;
+        }
+    }
+
+    if (tmr_freq != MCGIRCLK_HZ) {
+        config->bypassPrescaler = false;
+        config->value = prescaler;
+    } else {
+        config->bypassPrescaler = true;
+        config->value = 0;
+    }
+}
+
+static void
+timer_irq_handler(struct kinetis_hal_tmr *tmr)
+{
+    struct hal_timer *timer;
+    bool removed;
+
+    removed = false;
+    while ((timer = TAILQ_FIRST(&tmr->hal_timer_q)) != NULL) {
+        if ((int32_t)(kinetis_tmr_read(tmr) - timer->expiry) >= 0) {
+            TAILQ_REMOVE(&tmr->hal_timer_q, timer, link);
+            timer->link.tqe_prev = NULL;
+            timer->cb_func(timer->cb_arg);
+            removed = true;
+        } else {
+            break;
+        }
+    }
+
+    timer = TAILQ_FIRST(&tmr->hal_timer_q);
+    if (removed && timer) {
+        kinetis_tmr_set_period(tmr, timer->expiry);
+    }
+
+    tmr->overflow++;
```

Comment:
I don't think it will work as expected.
- If this was Free-Running mode (which I guess it is not) then overflow variable in the current form could hold upper 16 bits of 32 bits timer, but then it would need to be incremented not on every interrupt but only if overflow really occurred.
- It this is not Free-Running mode (I assume it is base on **LPTMR_GetDefaultConfig()** call), then overflow variable would not work as intended since counter overflows on different values (not every 0xFFFF but whenever CMR==CNR) which can be rather small. In this case reading timer will return some strange values.